rust/tw_evm/src/address.rs

```diff
@@ -0,0 +1,196 @@
+// Copyright © 2017-2023 Trust Wallet.
+//
+// This file is part of Trust. The full Trust copyright notice, including
+// terms governing use, modification, and redistribution, is contained in the
+// file LICENSE at the root of the source code distribution tree.
+
+use std::fmt::{Display, Formatter};
+use std::ops::{RangeFrom, RangeInclusive};
+use std::str::FromStr;
+use tw_coin_entry::coin_entry::CoinAddress;
+use tw_coin_entry::error::{AddressError, AddressResult};
+use tw_encoding::hex;
+use tw_hash::{sha3::keccak256, H160};
+use tw_keypair::ecdsa::secp256k1;
+use tw_memory::Data;
+
+pub trait EvmAddress: FromStr<Err = AddressError> + Into<Address> {
+    /// Tries to parse an address from the string representation.
+    /// Returns `Ok(None)` if the given `s` string is empty.
+    #[inline]
+    fn from_str_optional(s: &str) -> AddressResult<Option<Self>> {
+        if s.is_empty() {
+            return Ok(None);
+        }
+
+        Self::from_str(s).map(Some)
+    }
+}
+
+/// Represents an Ethereum address.
+#[derive(Clone, Copy, Debug, Default)]
+pub struct Address {
+    bytes: H160,
+}
+
+/// cbindgen:ignore
+impl Address {
+    pub const LEN: usize = 20;
+
+    /// Initializes an address with a `secp256k1` public key.
+    pub fn with_secp256k1_pubkey(pubkey: &secp256k1::PublicKey) -> Address {
+        const ADDRESS_HASH_RANGE: RangeFrom<usize> = 12..;
+
+        let pubkey_bytes = pubkey.uncompressed_without_prefix();
+        let hash = keccak256(pubkey_bytes.as_slice());
+        assert_eq!(hash.len(), 32);
+
+        let bytes = H160::try_from(&hash[ADDRESS_HASH_RANGE]).expect("Expected 20 byte array");
+
+        Address { bytes }
+    }
+
+    /// Displays the address in mixed-case checksum form
+    /// https://github.com/ethereum/EIPs/blob/master/EIPS/eip-55.md
+    fn into_checksum_address(self) -> String {
+        const UPPER_RANGE_1: RangeInclusive<char> = '8'..='9';
+        const UPPER_RANGE_2: RangeInclusive<char> = 'a'..='f';
+
+        let prefixed = false;
+        let addr_hex = hex::encode(self.bytes, prefixed);
+        let addr_hash = hex::encode(keccak256(addr_hex.as_bytes()), prefixed);
+
+        let payload_chars = addr_hex.chars().zip(addr_hash.chars()).map(|(a, h)| {
+            if a.is_ascii_digit() {
+                return a;
+            }
+            if UPPER_RANGE_1.contains(&h) || UPPER_RANGE_2.contains(&h) {
+                a.to_ascii_uppercase()
+            } else {
+                a.to_ascii_lowercase()
+            }
+        });
+
+        "0x".chars().chain(payload_chars).collect()
```

Comment:
```rust
let payload_chars = addr_hex.chars().zip(addr_hash.chars()).map(|(a, h)| {
        if a.is_ascii_digit() {
            a
        } else if (h <= UPPER_BOUND_1) || (h >= LOWER_BOUND_2 && h <= UPPER_BOUND_2) {
            a.to_ascii_uppercase()
        } else {
            a.to_ascii_lowercase()
        }
    });

    ["0x", &payload_chars.collect::<String>()].concat()
    ```
    this way? let's move the constant at module level?

Reply:
Done